Multimodality support

fellhorn opened the pull request:
Add bindings for the libmtmd multimodality support of llama-cpp. This allows to use models like e.g. Gemma3 with image or audio input. The bindings are only created with a new `mtmd` feature. To explain the usage, I added a simple example loosely based on the `mtmd-cli.cpp`.

To run the mtmd example, you first need to download the model gguf file and the multimodal projection file, e.g. for Gemma3 you may use:

```sh
wget https://huggingface.co/unsloth/gemma-3-4b-it-GGUF/resolve/main/gemma-3-4b-it-Q4_K_M.gguf \
https://huggingface.co/unsloth/gemma-3-4b-it-GGUF/resolve/main/mmproj-F16.gguf
```

To then run the example on CPU, provide an image file `my_image.jpg` and run:

```sh
cargo run --release --example mtmd -- \
  --model ./gemma-3-4b-it-Q4_K_M.gguf \
  --mmproj ./mmproj-F16.gguf \
  --image my_image.jpg \
  --prompt "What is in the picture?" \
  --no-gpu \
  --no-mmproj-offload \
  --marker "<start_of_image>"
```

Closes #744

MarcusDunn commented on llama-cpp-2/src/mtmd.rs:
```diff
@@ -0,0 +1,899 @@
+
+impl Default for MtmdContextParams {
+    fn default() -> Self {
+        Self {
+            use_gpu: false,
+            print_timings: true,
+            n_threads: 4,
+            media_marker: CString::new(mtmd_default_marker()).unwrap_or_default(),
+        }
+    }
```
this seems like it should rely on `mtmd_context_params_default`

MarcusDunn commented on llama-cpp-2/src/mtmd.rs:
```diff
@@ -0,0 +1,899 @@
+            )
+        };
+
+        if context.is_null() {
+            return Err(MtmdInitError::NullResult);
+        }
```
redundant check.

fellhorn replied:
Removed in [d025465](https://github.com/utilityai/llama-cpp-rs/pull/790/commits/d0254656235576803f9e41067d180381e37c89c5)

MarcusDunn commented on llama-cpp-2/src/mtmd.rs:
```diff
@@ -0,0 +1,899 @@
+        Ok(Self { context })
+    }
+
+    /// Check whether non-causal attention mask is needed before `llama_decode`.
+    #[must_use] pub fn decode_use_non_causal(&self) -> bool {
+        unsafe { llama_cpp_sys_2::mtmd_decode_use_non_causal(self.context.as_ptr()) }
+    }
```
a safety comment would be great. (self.context is not null should be sufficent?)

MarcusDunn commented on llama-cpp-2/src/mtmd.rs:
```diff
@@ -0,0 +1,899 @@
+    }
+}
+
+unsafe impl Send for MtmdContext {}
+unsafe impl Sync for MtmdContext {}
```
These impls need some sort of comment on why this is safe.

fellhorn replied:
Good catch, I actually meant `MtmdContextParams` and that is already `Send + Sync`.  Afais `mtmd_context` is not thread safe. Removed in d025465

MarcusDunn commented on llama-cpp-2/src/mtmd.rs:
```diff
@@ -0,0 +1,899 @@
+        if chunk_ptr.is_null() {
+            None
+        } else {
+            // Note: We don't own this chunk, it's owned by the chunks collection
```
should this return a reference then?

fellhorn commented on llama-cpp-2/src/mtmd.rs:
```diff
@@ -0,0 +1,899 @@
+            unsafe { CStr::from_ptr(ptr) }
+                .to_string_lossy()
+                .into_owned()
+                .into()
```
Cleaned it up in d025465

fellhorn commented:
@MarcusDunn, thanks for the thorough review. I think I addressed all your comments in d025465 and f149f11.
Especially your comment regarding feature gating everything related to mtmd should be addressed with f149f11:
I added another `wrapper_mtmd.h`, which will only be used when building with the mtmd feature.

Please let me know if there are still mtmd dependencies that I might have missed. I assume that the `include` in the Cargo.toml should be fine?

MarcusDunn commented on llama-cpp-2/src/mtmd.rs:
````diff
@@ -0,0 +1,931 @@
+/// ```
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum MtmdInputChunkType {
+    /// Text input chunk
+    Text = llama_cpp_sys_2::MTMD_INPUT_CHUNK_TYPE_TEXT as isize,
+    /// Image input chunk
+    Image = llama_cpp_sys_2::MTMD_INPUT_CHUNK_TYPE_IMAGE as isize,
+    /// Audio input chunk
+    Audio = llama_cpp_sys_2::MTMD_INPUT_CHUNK_TYPE_AUDIO as isize,
+}
````
is the cast required?

MarcusDunn reviewed (approved):
LGTM. Have the non-gemma models been tested? (I will merge regardless as we can always fix that later).

Feel free to address the comments, I'll merge a couple days from now to see if others want to give this branch some testing before I merge.

haixuanTao commented:
Can confirm the model mention above is now working wonders!

Thanks a lot!

haixuanTao commented:
More of an optimisation thing, but are we using the maximum number of batches for encoding images and audio? 

```rust
        self.n_past = chunks.eval_chunks(&self.mtmd_ctx, context, 0, 0, 1, true)?; // batch size of image encoding
```

It seems that this could be higher than 1, but not an expert on this

haixuanTao commented:
FYI, I tried to pass image+audio at the same time and it didn't work as there were only one marker within the prompt :D 

Not an urgent bug

haixuanTao commented:
> More of an optimisation thing, but are we using the maximum number of batches for encoding/decoding images and audio?
> 
> ```rust
>         self.n_past = chunks.eval_chunks(&self.mtmd_ctx, context, 0, 0, 1, true)?; // batch size of image decoding
> ```
> 
> It seems that this could be higher than 1, but not an expert on this
> 
> Related: [ggml-org/llama.cpp#14527](https://github.com/ggml-org/llama.cpp/issues/14527) if you search through the issue he's actually decoding at 64 n_tokens_batch

I am able to batch by 10 using qwen-omni, reducing the decoding time from 120ms-> 50ms, but I truly have no idea why most of time when giving a `n_batch` a random batch size it gives me the following error:

```bash
image slice encoded in 326 ms
decoding image batch 1/3, n_tokens_batch = 20
image decoded (batch 1/3) in 52 ms
decoding image batch 2/3, n_tokens_batch = 20
image decoded (batch 2/3) in 53 ms
decoding image batch 3/3, n_tokens_batch = 20
image decoded (batch 3/3) in 53 ms
get_logits_ith: invalid logits id 0, reason: batch.logits[0] != true
```

haixuanTao commented:
It's super fast in cuda. But it's a llama.cpp thing apparently

haixuanTao commented:
Related: https://github.com/ggml-org/llama.cpp/issues/15426

altunenes commented:
> Related: [ggml-org/llama.cpp#15426](https://github.com/ggml-org/llama.cpp/issues/15426)

 thanks. experiencing exactly the same issue. tried many things. I thought it was my fault. I've been working on it since yesterday. I tried running it on different threads, but no matter what I did, my optimization only improved it by 2-3 seconds at most. Also, I was doing all the image/video processing on the GPU side (using wgpu). 
Not tried Vulkan though
